torch/csrc/jit/passes/onnx/function_extraction.cpp

```diff
@@ -0,0 +1,1098 @@
+
+  while (IsValidScope(a) && IsValidScope(b)) {
+    if (a == b) {
+      return a;
+    } else {
+      a = a->parent();
+      b = b->parent();
+    }
+  }
+
+  return c10::nullopt;
+}
+
+c10::optional<ScopePtr> FunctionExtractor::FindCommonAncestor(
+    const scope_list& scopes) {
+  if (scopes.size() == 0) {
+    return c10::nullopt;
+  }
+
+  c10::optional<ScopePtr> common_ancestor = scopes.at(0);
+  for (auto scope : scopes) {
+    common_ancestor = FindCommonAncestor(common_ancestor.value(), scope);
+    if (!common_ancestor.has_value()) {
+      return c10::nullopt;
+    }
+  }
+
+  return common_ancestor;
+}
+
+c10::optional<ScopePtr> FunctionExtractor::InferScope(Node* n) {
+  // The scope of node n is assigned based on the following rules.
+  // 1. If all uses of outputs of n belongs to the same scope,
+  //    assign that scope, otherwise
+  // 2. If all nodes of inputs of n belongs to the same scope,
+  //    assign that scope, otherwise
+  // 3. Find common ancestor of the scopes of uses of outputs of n,
+  //    and the scopes of nodes of inputs of n.
+  scope_list input_scopes;
+  scope_list output_scopes;
+  for (auto input : n->inputs()) {
+    input_scopes.emplace_back(input->node()->scope());
+  }
+  for (auto output : n->outputs()) {
+    for (auto use : output->uses()) {
+      output_scopes.emplace_back(use.user->scope());
+    }
+  }
+  if (output_scopes.size() > 0 &&
+      std::all_of(
+          output_scopes.begin(),
+          output_scopes.end(),
+          [&output_scopes](ScopePtr scope) -> bool {
+            return IsValidScope(scope) && scope == output_scopes.at(0);
+          })) {
+    return output_scopes.at(0);
+  } else if (
+      input_scopes.size() > 0 &&
+      std::all_of(
+          input_scopes.begin(),
+          input_scopes.end(),
+          [&input_scopes](ScopePtr scope) -> bool {
+            return IsValidScope(scope) && scope == input_scopes.at(0);
+          })) {
+    return input_scopes.at(0);
+  } else {
+    scope_list scopes;
+    std::copy_if(
+        input_scopes.begin(), input_scopes.end(), scopes.begin(), IsValidScope);
+    std::copy_if(
+        output_scopes.begin(),
+        output_scopes.end(),
+        scopes.begin(),
+        IsValidScope);
+
+    if (scopes.size() > 0) {
+      // find common ancestor.
+      auto common_ancestor = FindCommonAncestor(scopes);
+
+      if (common_ancestor.has_value() &&
+          IsValidScope(common_ancestor.value())) {
+        return common_ancestor.value();
+      }
+    }
+  }
+
+  return c10::nullopt;
+}
+
+std::shared_ptr<Graph> FunctionExtractor::ConstructFuncGraph(
+    FunctionContext& func_ctx) {
+  auto& ctx = *func_ctx.scope_ctxs_[func_ctx.scope_key_];
+  const auto& nlist = ctx.nlist_;
+  const auto& scope = ctx.scope_;
+  auto& env = ctx.env_to_subgraph_;
+
+  auto g = std::make_shared<Graph>();
+  GRAPH_DEBUG("Constructing graph for ", scope->namesFromRoot());
+
+  // TODO: update input name of function to common name?
```

Comment:
Can you expand the comment? What is "common" in this context? Why would we want to do this TODO?